modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java

```diff
@@ -81,4 +87,44 @@ public static int findStartBuildingIndexCatalogVersion(CatalogService catalogSer
                 BUILDING, indexId, fromCatalogVersionIncluded, latestCatalogVersion
         ));
     }
+
+    /**
+     * Collects a list of tables that were removed from the catalog and should have been dropped due to a low watermark (if the catalog
+     * version in which the table was removed is less than or equal to the active catalog version of low watermark).
+     *
+     * @param catalogService Catalog service.
+     * @param lowWatermark Low watermark, {@code null} if it has never been updated.
+     * @return Result is sorted by the catalog version in which the table was removed from the catalog and by the table ID.
+     */
+    // TODO: IGNITE-21771 Process or check catalog compaction
+    static List<DroppedTableInfo> droppedTables(CatalogService catalogService, @Nullable HybridTimestamp lowWatermark) {
+        if (lowWatermark == null) {
+            return List.of();
+        }
+
+        int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+        int lwmCatalogVersion = catalogService.activeCatalogVersion(lowWatermark.longValue());
+
+        Set<Integer> previousCatalogVersionTableIds = Set.of();
+
+        var res = new ArrayList<DroppedTableInfo>();
+
+        for (int catalogVersion = earliestCatalogVersion; catalogVersion <= lwmCatalogVersion; catalogVersion++) {
+            int finalCatalogVersion = catalogVersion;
+
+            Set<Integer> tableIds = catalogService.tables(catalogVersion).stream()
+                    .map(CatalogObjectDescriptor::id)
+                    .collect(toSet());
+
+            difference(previousCatalogVersionTableIds, tableIds).stream()
+                    .map(tableId -> new DroppedTableInfo(tableId, finalCatalogVersion))
+                    .forEach(res::add);
+
+            previousCatalogVersionTableIds = tableIds;
+        }
+
+        res.sort(comparingInt(DroppedTableInfo::tableRemovalCatalogVersion).thenComparing(DroppedTableInfo::tableId));
```

Comment:
For the convenience of testing, so that the order of the tables is determined in a clear way, this sort will be called once and I think it will be quite fast.